SWC-453 Reference/Footnote Widget

heidiso opened the pull request.

heidiso committed:
Worked on the reference widget/syntax

heidiso committed:
New files create the reference widget

heidiso committed:
Edited the ConfigEditor for the reference widget. It no longer produces its own syntax. There is also only one field, the reference itself, for users to input right now.

heidiso committed:
Edited the test for the edited widget

heidiso committed:
SWC-453 Created a reference widget

The widget inserts anchors that redirects to the footnote at the end of the wiki.

heidiso committed:
SWC-453 Finished the reference widget

Made adjustments to the reference syntax to accept more formats.

heidiso committed:
Merge branch with ref widget implementation into develop

heidiso committed:
Fixed unneeded changes to the constants to avoid confusion.

heidiso committed:
Fixed unneeded changes to avoid confusion.

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/entity/SharedMarkdownUtils.java:
```diff
@@ -19,7 +19,13 @@ public static String getWidgetHTML(int widgetIndex, String suffix, String widget
-		sb.append("\" class=\"widgetContainer\" widgetParams=\"");
+		
+		//Reference widgets will be inline
+		if(widgetProperties.contains(WidgetConstants.REFERENCE_FOOTNOTE_KEY)) {
```
please change to a generic param that any widget could use to indicate that it should be placed inline.  The string that you're looking for might be something like "inlineWidget=true"

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/entity/registration/WidgetConstants.java:
```diff
@@ -95,7 +98,8 @@
 	public static final String BUTTON_LINK_URL_KEY = "url";
 	public static final String BUTTON_LINK_TEXT_KEY = "text";
-	
+	public static final String REFERENCE_TEXT_KEY = BUTTON_LINK_TEXT_KEY;
```
Just use BUTTON_LINK_TEXT_KEY (you can rename to something like BUTTON_TEXT_KEY)

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/entity/renderer/ReferenceWidgetViewImpl.java:
```diff
@@ -0,0 +1,62 @@
+		if(!hasLoaded) {
+			hasLoaded = true;
+			HTMLPanel parentPanel = (HTMLPanel)this.getParent();
+			Element heading = parentPanel.getElementById("footnote" + id);
```
please extract "footnote", and use in the parser

jay-hodgson commented on src/main/java/org/sagebionetworks/web/server/markdownparser/ReferenceParser.java:
```diff
@@ -0,0 +1,43 @@
+		StringBuffer sb = new StringBuffer();
+		while(m.find()) {
+			//Store reference text and insert extra parameter for the renderer to link to
+			footnotes.add(input.substring(m.start() + 17, m.end() - 1));
```
Please make where 17 came from obvious.  like "${reference?text=".length() (at least I think that's where this comes from)

jay-hodgson commented on src/main/java/org/sagebionetworks/web/server/markdownparser/ReferenceParser.java:
```diff
@@ -0,0 +1,43 @@
+	@Override
+	public void completeParse(StringBuilder html) {
+		for(int i = 0; i < footnotes.size(); i++) {
+			String text = footnotes.get(i).replaceAll("%2E", ".");
```
This might be a bigger issue with over-escaping

heidiso committed:
SWC-453 Last adjustments to the reference widget

*Fixed the bug that made characters (following widget syntax) at the end of lines disappear.

heidiso committed:
Merge branch 'issue453_ref_widget' into develop

jay-hodgson commented on src/main/java/org/sagebionetworks/web/server/markdownparser/ReferenceParser.java:
```diff
@@ -0,0 +1,49 @@
+			int lenOfSyntax = "${reference?text=".length();
+			footnotes.add(input.substring(m.start() + lenOfSyntax, m.end() - 1));
+			String updated = input.substring(m.start(), m.end() - 1) + 
+				"&" + WidgetConstants.INLINE_WIDGET_KEY + "=true&" + WidgetConstants.REFERENCE_FOOTNOTE_KEY + "=" + footnoteNumber + "}";
```
Please put(WidgetConstants.INLINE_WIDGET_KEY, true) in the widget params from the editor instead of injecting in the parser.  Other widgets should do the same thing without having a parser side.

heidiso committed:
SWC-453 Added description on how to make widgets inline

heidiso committed:
SWC-453 Now adding the inline parameter in the editor rather than the parser
